- replaced custom BookRepository with JpaRepository interface

DanilKi opened the pull request:
- implemented soft delete concept for the Book entity
- added update & delete by id methods for Book entity to BookController & BookServiceImpl classes
- added liquibase support

DanilKi committed:
- replaced custom BookRepository with JpaRepository interface
- implemented soft delete concept for the Book entity
- added update & delete by id methods for Book entity to BookController & BookServiceImpl classes
- added liquibase support

Elena-Bruyako commented on src/main/java/com/onlinebookstore/model/Book.java:
```diff
@@ -7,9 +7,13 @@
 @Entity
 @Data
+@SQLDelete(sql = "UPDATE book SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
```
Add Table annotation with name

Elena-Bruyako replied:
Use `@SQLRestriction` instead of `@Where`

Elena-Bruyako commented on src/main/java/com/onlinebookstore/model/Book.java:
```diff
@@ -7,9 +7,13 @@
+import org.hibernate.annotations.Where;
 
 @Entity
 @Data
```
Don't use `@Data` for entities, read article about it

Elena-Bruyako commented on src/main/java/com/onlinebookstore/service/impl/BookServiceImpl.java:
```diff
@@ -23,6 +23,20 @@ public BookDto save(CreateBookRequestDto requestDto) {
+        Book book = bookRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't update a book by id: " + id)
+        );
+        book.setTitle(requestDto.getTitle());
```
It looks like mapper can implement this logic

Elena-Bruyako commented on src/main/resources/db/changelog/changes/01-create-book-table.yml:
```diff
@@ -0,0 +1,48 @@
+      author: DanilKi
+      changes:
+        - createTable:
+            tableName: book
```
Table name should be plural
```suggestion
            tableName: books
```

Elena-Bruyako commented on src/main/java/com/onlinebookstore/controller/BookController.java:
```diff
@@ -5,11 +5,15 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
```
Let's add `server.servlet.context-path=/api` into application.properties file:
`@RequestMapping("/books")`

Elena-Bruyako reviewed (changes requested):
Good job! Let's improve your solution!

DanilKi committed:
changes for Book entity:
  - added @Table annotation
  - replaced @Data with @Getter @Setter annotation
  - replaced deprecated @Where annotation with @SQLRestriction
  - renamed table 'book' to 'books' in all sources & config

changes for Book handling logic:
  - moved mapping logic from BookServiceImpl.update() method to mapper
  - added overloaded toBookEntity() method to BookMapper interface

DanilKi commented:
I have a question regarding the update operation. Is there any way to perform an UPDATE without performing a SELECT before it, using JpaRepository? I tried several variants, but Hibernate fetches entity from DB by select stmt, before update statement (I saw that in console) or I get LazyInitEx. I mean, maybe there are some standard methods exist, without declaring custom method with @Query annotation and writing SQL statement in it?

Elena-Bruyako commented on src/main/java/com/onlinebookstore/mapper/BookMapper.java:
```diff
@@ -5,10 +5,13 @@
 
     Book toBookEntity(CreateBookRequestDto requestDto);
+
+    Book toBookEntity(CreateBookRequestDto requestDto, @MappingTarget Book book);
```
```suggestion
    void toBookEntity(CreateBookRequestDto requestDto, @MappingTarget Book book);
```

Elena-Bruyako commented on src/main/java/com/onlinebookstore/service/impl/BookServiceImpl.java:
```diff
@@ -23,6 +23,15 @@ public BookDto save(CreateBookRequestDto requestDto) {
+        Book book = bookRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't update a book by id: " + id)
+        );
+        book = bookMapper.toBookEntity(requestDto, book);
```
```suggestion
        bookMapper.toBookEntity(requestDto, book);
```

Elena-Bruyako reviewed (approved):
Minor comment
let's discuss your question in chat

DanilKi committed:
made toBookEntity(requestDto, Book) void

Co-authored-by: Olena Bruiako <bruyako.elena@gmail.com>
